src/project/model.rs

```diff
@@ -0,0 +1,121 @@
+use cairo_lang_filesystem::ids::CrateLongId;
+use std::collections::{HashMap, HashSet};
+use std::path::{Path, PathBuf};
+
+use crate::lang::db::AnalysisDatabase;
+use crate::lang::proc_macros::controller::ProcMacroClientController;
+use crate::project::{Crate, ManifestRegistry};
+
+#[derive(Default)]
+pub struct ProjectModel {
+    // The two fields below keep exactly the same information;
+    // therefore, their contents should be kept synchronised.
+    // We keep both of them for efficiency and ease of use.
+    loaded_workspaces: HashMap<PathBuf, HashMap<CrateLongId, Crate>>,
+    loaded_crates: HashMap<CrateLongId, HashSet<PathBuf>>,
+}
+
+impl ProjectModel {
+    pub fn clear_loaded_workspaces(&mut self) {
+        self.loaded_workspaces.clear();
+        self.loaded_crates.clear();
+    }
+
+    pub fn load_workspace(
+        &mut self,
+        db: &mut AnalysisDatabase,
+        workspace_crates: Vec<Crate>,
+        workspace_dir: PathBuf,
+        proc_macro_controller: &ProcMacroClientController,
+        manifest_registry: &ManifestRegistry,
+        enable_linter: bool,
+    ) {
+        let workspace_crates = workspace_crates.into_iter().map(|cr| (cr.long_id(), cr)).collect();
+        if let Some(old_crates) = self.loaded_workspaces.get(&workspace_dir) {
+            if old_crates == &workspace_crates {
+                return;
+            }
+
+            // Static because the borrow checker is stupid.
+            ProjectModel::remove_crates(&mut self.loaded_crates, &workspace_dir, old_crates);
+        };
+
+        self.add_crates(workspace_crates, &workspace_dir);
+
+        self.apply_changes_to_db(db, proc_macro_controller, manifest_registry, enable_linter);
+    }
+
+    fn remove_crates(
+        loaded_crates: &mut HashMap<CrateLongId, HashSet<PathBuf>>,
+        workspace_dir: &PathBuf,
```

Comment:
```suggestion
        workspace_dir: &Path,
```

it is like changing `&String` to `&str`